src/validation.h

```diff
@@ -194,12 +194,22 @@ void UnlinkPrunedFiles(const std::set<int>& setFilesToPrune);
 /** Prune block files up to a given height */
 void PruneBlockFilesManual(int nManualPruneHeight);
 
-/** (try to) add transaction to memory pool
- * plTxnReplaced will be appended to with all transactions replaced from mempool
- * @param[out] fee_out optional argument to return tx fee to the caller **/
-bool AcceptToMemoryPool(CTxMemPool& pool, TxValidationState &state, const CTransactionRef &tx,
-                        std::list<CTransactionRef>* plTxnReplaced,
-                        bool bypass_limits, bool test_accept=false, CAmount* fee_out=nullptr) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
+/** Per-transaction result from trying to accept a transaction to the memory pool. */
+struct MempoolAcceptResult {
+    MempoolAcceptResult(const CTransactionRef& ptx) : txid(ptx->GetHash()) {}
+
+    uint256 txid;
+    bool m_accepted = false;
+    TxValidationState m_state;
+    std::list<CTransactionRef> m_replaced_transactions{};
+    CAmount m_fee = CAmount(0);
```

Comment:
Not sure if it makes sense to return a fee of `0` when the tx was rejected. At the very least the members that are optional, should be `std::optional`. Though, I am thinking that it could make sense to return two completely different types, based on whether the tx was accepted? I.e. a struct with `{fee, replaced_txs}` if the tx was accepted and a struct with `{state}` if the tx was rejected.

This would make it harder at the call sites to make mistakes such as returning the fee of a rejected tx.

Reply:
Ah true! I'm also starting to think that `m_accepted == m_state.IsValid()` always?? 🤔 maybe could get rid of that and do

```c
TxValidationState m_state;
std::optional<std::list<CTransactionRef>> m_replaced_transactions;
std::optional<CAmount> m_base_fee;
```

Reply:
Not sure about the equivalence right now, but `if m_accepted then m_state.IsValid()` should hold.

Reply:
Yeah I agree - I just started writing this and I think we should actually put the `TxValidationState` and `m_base_fees` into the `MemPoolAccept::Workspace` (since we need access throughout validation), then return whichever is appropriate at the end.

How does this look? Kinda Rustic:
```c

struct MempoolAcceptOk {
    std::list<CTransactionRef> m_replaced_transactions;
    CAmount m_base_fee;
    // Other stuff
}

struct MempoolAcceptError {
    uint256 txid; // otherwise we don't know which tx failed
    TxValidationState m_state;
}

// either return a list of MempoolAcceptOks or 1 MempoolAcceptError
// function signature for ATMP and ProcessNewPackage would thus be the same
using MempoolAcceptResult = std::variant<std::vector<MempoolAcceptOk>, MempoolAcceptError>;
```

Reply:
Looks good

Reply:
Oops. I just realized `Workspace` is intended to be const after `PreChecks` populates it, so it can't hold `TxValidationState` since that can mutate during the script checks.
Will do the `std::optional` approach for now, it's actually a bit cleaner I think.

Reply:
ay @MarcoFalke, the `std::variant` method is a bit hairy... see [this branch](https://github.com/glozow/bitcoin/commit/7d9151afc0d95e232bf1f57e8e052f5dc3bef7e7). It works, but has a larger diff so I'm worried it might encumber reviewers. If you think it's an improvement, I can leave it for a followup or something?